#1 feature(WpfControls): Выполнено дополнительное задание по ООАиПр

Melaved opened the pull request.

Melaved committed:
featrure(WpfControls): Программа доделана
- реализована верстка
- реализована сериализация данных, чтобы файлы в списке сохранялись между сессиями

inittose commented on Model/FileItem.cs:
```diff
@@ -2,19 +2,25 @@
+        /// Полный путь к файлу.
         /// </summary>
         [ObservableProperty]
         private string _filePath;
```
`_filePath` заменить на `_path`, т.к. в названии класса уже содержится упоминание File (`FileItem`), поэтому явно понятно, что это путь к файлу

Melaved replied:
если назвать переменную _path, то будет ошибка
![image](https://github.com/user-attachments/assets/e2d910d3-9998-4806-a37b-ddbf59404b23)
так как появится неоднозначность между свойством Path и Path из System.IO

inittose replied:
Хорошо, оставь `_filePath`, либо можно записать, например:
```c#
public string Name => System.IO.Path.GetFileName(Path)
```
На твое усмотрение

inittose commented on Model/FileItem.cs:
```diff
@@ -2,19 +2,25 @@
 
         /// <summary>
-        /// Возвращает имя файла.
+        /// Получает имя файла (без пути).
```
**Получает** заменить на **Возвращает**
Убрать "(без пути)", т.к. названние свойства подразумевает, что возвращается только имя

inittose commented on ViewModel/MainViewModel.cs:
```diff
@@ -2,15 +2,28 @@
     {
-        public ObservableCollection<FileItem> Files { get; } = new();
+        /// <summary>
+        /// Коллекция файлов, отображаемая в UI.
+        /// Инициализируется данными из файла при создании ViewModel.
```
Возвращает коллекцию файлов
То что файлы отображаются в UI можно не упоминать

inittose commented on WpfControls/MainWindow.xaml.cs:
```diff
@@ -4,16 +4,37 @@
+        /// <param name="e">Данные события <see cref="RoutedEventArgs"/>.</param>
+        private void AddFiles_Click(object sender, RoutedEventArgs e)
+        {
+            var dialog = new Microsoft.Win32.OpenFileDialog
```
`dialog` стоит заменить на `fileDialog`

inittose commented on WpfControls/MainWindow.xaml.cs:
```diff
@@ -4,16 +4,37 @@
+                Filter = "All files (*.*)|*.*"
+            };
+
+            if (dialog.ShowDialog() != true || DataContext is not MainViewModel vm)
```
`vm` - необщепринятое сокращение
Заменить на `viewModel`

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,110 @@
+                                    <TextBlock
+                                        Margin="5 2"
+                                        VerticalAlignment="Center"
+                                        Text="{Binding Name}"
+                                        TextTrimming="CharacterEllipsis"/>
+
+                                    <!--  Кнопка удаления  -->
+                                    <Button
+                                        Grid.Column="1"
+                                        Margin="5 0 5 0"
+                                        VerticalAlignment="Center"
+                                        Command="{Binding DataContext.RemoveFileCommand, RelativeSource={RelativeSource AncestorType=Window}}"
+                                        CommandParameter="{Binding}"
+                                        Content="✖"
+                                        Style="{StaticResource FlatButtonStyle}">
+                                        <Button.Visibility>
+                                            <Binding
+                                                Converter="{StaticResource BoolToVisibilityConverter}"
+                                                Path="IsMouseOver"
+                                                RelativeSource="{RelativeSource AncestorType=ListBoxItem}"/>
+                                        </Button.Visibility>
+                                    </Button>
+                                </Grid>
+                            </Border>
+                        </DataTemplate>
+                    </ListBox.ItemTemplate>
+                </ListBox>
+
+                <!--  Кнопка добавления  -->
+                <Button
+                    Width="35"
+                    Height="35"
+                    Margin="10"
+                    HorizontalAlignment="Left"
+                    Click="AddFiles_Click">
+                    <Image Source="Resources/Add.png" Stretch="Uniform"/>
+                </Button>
+            </StackPanel>
         </ScrollViewer>
     </Grid>
```
Сейчас верстка выглядит вот так:
![изображение](https://github.com/user-attachments/assets/e6123e2e-088c-4f0e-8753-87008fdb6bc5)
Но желательно, чтобы выглядела следующим образом:
![изображение](https://github.com/user-attachments/assets/fbcf7d72-9278-459a-a5ce-65a184bba192)
То есть необходимо, чтобы кнопка "добавить" всегда находилась снизу, вне зависимости от количества элементов в ListBox

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,110 @@
+                                        Margin="5 2"
+                                        VerticalAlignment="Center"
+                                        Text="{Binding Name}"
+                                        TextTrimming="CharacterEllipsis"/>
```
Свойство `TextTrimming` не будет работать, так как `ListBox` содержит `ScrollViewer` в шаблоне. Необходимо задать `ListBox`'у такое свойство, чтобы `ScrollViewer` не мог расширять `ListBoxItem`


![изображение](https://github.com/user-attachments/assets/05de7e98-6080-4d65-9d72-6fc2d7018d7b)
\\/
![изображение](https://github.com/user-attachments/assets/d35f8afb-f79b-4925-bb02-e4ab85b6e47b)
Кнопка удаления тоже всегда должна быть видна и не зависеть от количества символов в имени файла

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,110 @@
+                </ListBox>
+
+                <!--  Кнопка добавления  -->
+                <Button
+                    Width="35"
+                    Height="35"
+                    Margin="10"
+                    HorizontalAlignment="Left"
+                    Click="AddFiles_Click">
+                    <Image Source="Resources/Add.png" Stretch="Uniform"/>
+                </Button>
```
Кнопкам добавить стиль, чтобы фон и граница были прозрачными при наведении мыши и нажатии на кнопки
Также при:
- Наведении курсора `Opacity = 0.5`
- Нажатии кнопки `Opacity = 1`
- Для неактивного состояния `Opacity = 0.75`

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,110 @@
+        <!--  Вся область с прокруткой  -->
+        <ScrollViewer VerticalScrollBarVisibility="Auto">
+            <StackPanel>
+                <ListBox
+                    HorizontalContentAlignment="Stretch"
+                    Background="Transparent"
+                    BorderThickness="0"
+                    ItemsSource="{Binding Files}">
```
Добавить стиль для `ListBoxItem`, чтобы при наведении и нажатии на элемент фон и граница не менялись, также, можно в стиль для `ListBoxItem` установить фон для невалидных файлов, и использовать его вместо `Border`
(через свойство `ItemContainerStyle` у `ListBox` можно задать стиль `ListBoxItem`)

inittose reviewed (changes requested):
Если возникнут проблемы со стилями или версткой, задавай вопросы
Также советую посещать официальный репозиторий [wpf](https://github.com/dotnet/wpf/tree/main), чтобы получить больше информации о стилях элементов управления

Melaved committed:
style(WpfControls): Исправлены замечания
- исправлены комментарии
- переменованы переменные

Melaved committed:
feature(View): Исправлены замечания
- кнопка "добавить" теперь находится внизу вне зависимости от количества элементов
- имена файлов обрезаются правильно по длине окна
- добавлен стиль для кнопок чтобы фон и граница были прозрачными при наведении мыши и нажатии на кнопки

Melaved committed:
style(View): отформатировал XAML

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,147 @@
+        <BooleanToVisibilityConverter x:Key="BoolToVisibilityConverter"/>
+
+        <!--  Стиль для кнопок  -->
+        <Style x:Key="ButtonStyle" TargetType="Button">
+            <Setter Property="Background" Value="Transparent"/>
+            <Setter Property="BorderBrush" Value="Transparent"/>
+            <Setter Property="BorderThickness" Value="0"/>
+            <Setter Property="Cursor" Value="Hand"/>
+            <Style.Triggers>
+                <Trigger Property="IsMouseOver" Value="True">
+                    <Setter Property="Opacity" Value="0.5"/>
+                </Trigger>
+                <Trigger Property="IsPressed" Value="True">
+                    <Setter Property="Opacity" Value="1"/>
+                </Trigger>
+                <Trigger Property="IsEnabled" Value="True">
+                    <Setter Property="Opacity" Value="0.75"/>
+                </Trigger>
+            </Style.Triggers>
+        </Style>
```
Из-за того, что в стандартном шаблоне кнопки прописаны триггеры, то их надо переопределить, чтобы они работали
Пересмотрел поведение, пусть при наведении на кнопку Opacity = 0.75
Ниже я привел пример, как можно переопределить шаблон
```c#
<Style x:Key="ButtonStyle" TargetType="Button">
    <Setter Property="Background" Value="Transparent"/>
    <Setter Property="BorderBrush" Value="Transparent"/>
    <Setter Property="BorderThickness" Value="0"/>
    <Setter Property="Cursor" Value="Hand"/>
    <Setter Property="Template">
        <Setter.Value>
            <ControlTemplate TargetType="Button">
                <Border
                    x:Name="border"
                    Background="{TemplateBinding Background}"
                    BorderBrush="{TemplateBinding BorderBrush}"
                    BorderThickness="{TemplateBinding BorderThickness}"
                    CornerRadius="5">
                    <ContentPresenter HorizontalAlignment="Center" VerticalAlignment="Center"/>
                </Border>
                <ControlTemplate.Triggers>
                    <Trigger Property="IsMouseOver" Value="True">
                        <Setter TargetName="border" Property="Opacity" Value="0.75"/>
                    </Trigger>
                    <Trigger Property="IsPressed" Value="True">
                        <Setter TargetName="border" Property="Opacity" Value="1"/>
                    </Trigger>
                </ControlTemplate.Triggers>
            </ControlTemplate>
        </Setter.Value>
    </Setter>
</Style>
```

inittose commented on WpfControls/MainWindow.xaml:
```diff
@@ -3,37 +3,147 @@
+        </Style>
+
+        <!--  Стиль кнопки удаления  -->
+        <Style x:Key="FlatButtonStyle" TargetType="Button">
```
В стилях WPF есть концепция наследования стилей
Отнаследуйся от стиля `ButtonStyle`, чтобы кнопка удаления имела такое же поведение наведения на кнопку, нажатия

Melaved committed:
refactor(View): Исправлены замечания
- переопределил триггеры и изменил Opacity при наведении для ButtonStyle
- отнаследовал стиль кнопки удаления(FlatButtonStyle) от ButtonStyle
